lib/core/CLoggerThrottler.cc

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CLoggerThrottler.h>
+
+#include <core/CTimeUtils.h>
+
+#include <limits>
+#include <iostream>
+
+namespace ml {
+namespace core {
+
+CLoggerThrottler& CLoggerThrottler::instance() {
+    return ms_Instance;
+}
+
+void CLoggerThrottler::minimumLogInterval(std::int64_t minimumLogInterval) {
+    m_MinimumLogInterval = minimumLogInterval;
+}
+
+bool CLoggerThrottler::skip(const char* file, int line) {
+
+    auto key = std::make_pair(file, line);
+    auto now = CTimeUtils::nowMs();
+
+    std::unique_lock<std::mutex> lock{m_Mutex};
+    auto lastTime = m_LastLogTimes.emplace(key, std::numeric_limits<std::int64_t>::min()).first;
+
+    if (now > lastTime->second + m_MinimumLogInterval) {
+        lastTime->second = now;
+        return false;
+    }
+    return true;
+}
+
+CLoggerThrottler::CLoggerThrottler() : m_MinimumLogInterval{3600 * 1000} {}
+
+CLoggerThrottler CLoggerThrottler::ms_Instance;
```

Comment:
Code that runs during static initialisation might contain log messages, so I think this will have to be constructed using a pattern that isn't vulnerable to the order of static initialisation, for example:

```
CLoggerThrottler& CLoggerThrottler::instance() {
    static CLoggerThrottler instance;
    return instance;
}
```

And then get rid of the `ms_Instance` variable.

The other way to do it would be to make this class not a singleton but a non-static member of the `CLogger` class, so that if the `CLogger` singleton exists then it's guaranteed that its corresponding `CLoggerThrottler` also does.  So the macros would call `ml::core::CLogger::instance().throttler()` similarly to how they currently call `ml::core::CLogger::instance().logger()` to get the low level logger object.

Reply:
I like the idea of this being a member of the logger class see [6c93df4](https://github.com/elastic/ml-cpp/pull/1615/commits/6c93df43771cced3678e08982ffc89508e46806f).

(I don't think we would currently suffer from static initialisation order issues, because I don't think we do anything complicated enough at static variable initialisation time, but do agree we should be defensive to this case.)

Reply:
> (I don't think we would currently suffer from static initialisation order issues, because I don't think we do anything complicated enough at static variable initialisation time, but do agree we should be defensive to this case.)

There is currently one place: https://github.com/elastic/ml-cpp/blob/685b55cfd04b9490ff923417801a272ab148ff43/lib/core/CTimezone_Windows.cc#L54-L55

We used to have more of these, which is the origin of the `DO_NOT_USE_THIS_VARIABLE` variables.  To solve the static initialization order problem we had to use the Meyers singleton pattern for our singletons, but then in the days where function-local static initialization wasn't thread safe we had to ensure they were initialized before `main()` ran.  (In Visual Studio 2015 function-local static initialization was finally made thread safe, so all the `DO_NOT_USE_THIS_VARIABLE` variables can be deleted now - I will do that in a separate PR once this is merged.)